---
title: "Example Plots"
author: "Mark Dunning"
date: "15 November 2017"
output: html_document
---


## Gene Expression Data

We are going to get some public data from the Bioconductor website which in included in the `breastCancerNKI` package. This code will download the package if it is not already installed.

```{r eval=FALSE}

source("http://www.bioconductor.org/biocLite.R")
biocLite("breastCancerNKI",suppressUpdates = TRUE)
biocLite("genefilter",suppressUpdates = TRUE)
biocLite("pheatmap",suppressUpdates = TRUE)  
```

The `exprs` function can be used to extract the expression values for all patients in the study and all genes. The `rowSds` function will calculate the standard deviation for each gene

```{r}
library(breastCancerNKI)
library(Biobase)
library(genefilter)
library(pheatmap)
library(RColorBrewer)

data(nki)

expression_matrix <- exprs(nki)
features <- fData(nki)
patient_data <- pData(nki)

gene_var <- rowSds(expression_matrix)
top50 <- order(gene_var,decreasing = TRUE)[1:50]

Grade <- factor(patient_data$grade)
levels(Grade) <- c("Low","Medium","High")

sample_anno <- data.frame(ER = ifelse(patient_data$er==0,"Negative","Positive"),Grade)
rownames(sample_anno) <- colnames(expression_matrix)

ann_colors <- list(
  ER = c(Negative = "#E41A1C", Positive = "#377EB8"),
  Grade = c(Low = "#4DAF4A", Medium = "#984EA3",  High = "#FF7F00")
)


    

pheatmap(expression_matrix[top50,],annotation_col = sample_anno,
         annotation_colors = ann_colors,
         labels_row = features$NCBI.gene.symbol[top50],
         main = "Gene Expression of 50 most variable genes")
```


```{r}
png("GeneExpressionHeatmap.png",width=800,height=800)
par(mfrow=c(1,2))
pheatmap(expression_matrix[top50,],annotation_col = sample_anno,
         annotation_colors = ann_colors,
         labels_row = features$NCBI.gene.symbol[top50],
         main = "Gene Expression of 50 most variable genes")
dev.off()
```


```{r}
png("SurvivalCurves.png",width=400,height=400)
library(survival)
plot (survfit (Surv(pData(nki)$t.dmfs , pData(nki)$e.dmfs) ~
  pData(nki)$er) , col = c("#E41A1C" , "#377EB8"),lwd=10,
    xlab="Survival Time",
    ylab="Proportion of Patients")
dev.off()
```

## ggbio

```{r eval=FALSE}
biocLite("ggbio",suppressUpdates=TRUE)
```




```{r}
library(ggbio)
snp <- read.table(system.file("extdata", "plink.assoc.sub.txt", package = "biovizBase"),
header = TRUE)
require(biovizBase)
gr.snp <- transformDfToGr(snp, seqnames = "CHR", start = "BP", width = 1)
head(gr.snp)
## change the seqname order
require(GenomicRanges)
gr.snp <- keepSeqlevels(gr.snp, as.character(1:22))
seqlengths(gr.snp)
## need to assign seqlengths
data(ideoCyto, package = "biovizBase")
seqlengths(gr.snp) <- as.numeric(seqlengths(ideoCyto$hg18)[1:22])
## remove missing
gr.snp <- gr.snp[!is.na(gr.snp$P)]
## transform pvalue
values(gr.snp)$pvalue <- -log10(values(gr.snp)$P)
head(gr.snp)

plotGrandLinear(gr.snp, aes(y = pvalue), color = c("#7fc97f", "#fdc086"),
cutoff = 3, cutoff.color = "blue", cutoff.size = 0.2)
```


```{r}
data(darned_hg19_subset500, package = "biovizBase")
dn <- darned_hg19_subset500
library(GenomicRanges)
seqlengths(dn)
## add seqlengths
## we have seqlegnths information in another data set
seqlengths(dn) <- seqlengths(ideoCyto$hg19)[names(seqlengths(dn))]
## then we change order
dn <- keepSeqlevels(dn, paste0("chr", c(1:22, "X")))
autoplot(dn, layout = "karyogram")

autoplot(dn, layout = "karyogram", aes(color = exReg, fill = exReg), alpha = 0.5) +
scale_color_discrete(na.value = "brown")
```

```{r}
data("CRC", package = "biovizBase")
p <- ggbio(trackWidth = 10, buffer = 0, radius = 10) + circle(hg19sub, geom = "ideo", fill = "gray70") +
circle(hg19sub, geom = "scale", size = 2) +
circle(hg19sub, geom = "text", aes(label = seqnames), vjust = 0, size = 3)
p


grl <- split(crc.gr, values(crc.gr)$individual)
## need "unit", load grid
library(grid)
crc.lst <- lapply(grl, function(gr.cur){
print(unique(as.character(values(gr.cur)$individual)))
cols <- RColorBrewer::brewer.pal(3, "Set2")[2:1]

names(cols) <- c("interchromosomal", "intrachromosomal")
p <- ggbio() + circle(gr.cur, geom = "link", linked.to = "to.gr",
aes(color = rearrangements)) +
circle(hg19sub, geom = "ideo",
color = "gray70", fill = "gray70") +
scale_color_manual(values = cols) +
labs(title = (unique(values(gr.cur)$individual))) +
theme(plot.margin = unit(rep(0, 4), "lines"))
})

arrangeGrobByParsingLegend(crc.lst, widths = c(4, 1), legend.idx = 1, ncol = 3)
```



## Oncoprint

```{r eval=FALSE}
biocLite("ComplexHeatmap",suppressUpdates = TRUE)
```


https://www.bioconductor.org/packages/devel/bioc/vignettes/ComplexHeatmap/inst/doc/s8.oncoprint.html#toc_2


```{r}
mat = read.table(paste0(system.file("extdata", package = "ComplexHeatmap"), 
    "/tcga_lung_adenocarcinoma_provisional_ras_raf_mek_jnk_signalling.txt"), 
    header = TRUE,stringsAsFactors=FALSE, sep = "\t")
mat[is.na(mat)] = ""
rownames(mat) = mat[, 1]
mat = mat[, -1]
mat=  mat[, -ncol(mat)]
mat = t(as.matrix(mat))
mat[1:3, 1:3]
```

```{r}
alter_fun = list(
    background = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0.5, "mm"), h-unit(0.5, "mm"), gp = gpar(fill = "#CCCCCC", col = NA))
    },
    HOMDEL = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0.5, "mm"), h-unit(0.5, "mm"), gp = gpar(fill = "blue", col = NA))
    },
    AMP = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0.5, "mm"), h-unit(0.5, "mm"), gp = gpar(fill = "red", col = NA))
    },
    MUT = function(x, y, w, h) {
        grid.rect(x, y, w-unit(0.5, "mm"), h*0.33, gp = gpar(fill = "#008000", col = NA))
    }
)
col = c("MUT" = "#008000", "AMP" = "red", "HOMDEL" = "blue")
```

```{r}
library(ComplexHeatmap)
png("oncoprint.png",width=800,height=800)
oncoPrint(mat, get_type = function(x) strsplit(x, ";")[[1]],
    alter_fun = alter_fun, col = col, 
    column_title = "OncoPrint for TCGA Lung Adenocarcinoma, genes in Ras Raf MEK JNK signalling",
    heatmap_legend_param = list(title = "Alternations", at = c("AMP", "HOMDEL", "MUT"), 
        labels = c("Amplification", "Deep deletion", "Mutation")))
dev.off()
```

## Pathview

```{r eval=FALSE}

source("http://www.bioconductor.org/biocLite.R")
biocLite("pathview",suppressUpdates = FALSE)

```

```{r}
library(pathview)
```


## t-SNE

```{r eval=FALSE}
biocLite("SingleCellExperiment",suppressUpdates = TRUE)
biocLite("scater",suppressUpdates = TRUE)
biocLite("mvoutlier",suppressUpdates = TRUE)
```


```{r}
download.file("https://github.com/hemberg-lab/scRNA.seq.course/raw/master/tung/molecules.txt",destfile = "molecules.txt")
download.file("https://raw.githubusercontent.com/hemberg-lab/scRNA.seq.course/master/tung/annotation.txt",destfile = "annotation.txt")
```

```{r}

```






















```{r}
download.file("https://github.com/hemberg-lab/scRNA.seq.course/raw/master/tung/reads.txt",destfile = "reads.txt")

library(SingleCellExperiment)
library(scater)

reads <- read.table("reads.txt", sep = "\t")
anno <- read.table("annotation.txt", sep = "\t", header = TRUE)

reads <- SingleCellExperiment(
    assays = list(counts = as.matrix(reads)), 
    colData = anno
)

```

```{r}
keep_feature <- rowSums(counts(reads) > 0) > 0
reads <- reads[keep_feature, ]
```

```{r}
isSpike(reads, "ERCC") <- grepl("^ERCC-", rownames(reads))
isSpike(reads, "MT") <- rownames(reads) %in% 
    c("ENSG00000198899", "ENSG00000198727", "ENSG00000198888",
    "ENSG00000198886", "ENSG00000212907", "ENSG00000198786",
    "ENSG00000198695", "ENSG00000198712", "ENSG00000198804",
    "ENSG00000198763", "ENSG00000228253", "ENSG00000198938",
    "ENSG00000198840")
```

```{r}
reads <- calculateQCMetrics(
    reads,
    feature_controls = list(
        ERCC = isSpike(reads, "ERCC"), 
        MT = isSpike(reads, "MT")
    )
)
```

```{r}
filter_by_total_counts <- (reads$total_counts > 1.3e6)
filter_by_expr_features <- (reads$total_features > 7000)
filter_by_ERCC <- 
    reads$batch != "NA19098.r2" & reads$pct_counts_ERCC < 25
filter_by_MT <- reads$pct_counts_MT < 30
reads$use <- (
    # sufficient features (genes)
    filter_by_expr_features &
    # sufficient molecules counted
    filter_by_total_counts &
    # sufficient endogenous RNA
    filter_by_ERCC &
    # remove cells with unusual number of reads in MT genes
    filter_by_MT
)

reads <- plotPCA(
    reads,
    size_by = "total_features", 
    shape_by = "use",
    pca_data_input = "pdata",
    detect_outliers = TRUE,
    return_SCE = TRUE
)

filter_genes <- apply(
    counts(reads[, colData(reads)$use]), 
    1, 
    function(x) length(x[x > 1]) >= 2
)
rowData(reads)$use <- filter_genes
assay(reads, "logcounts_raw") <- log2(counts(reads) + 1)
reducedDim(reads) <- NULL
saveRDS(reads, file = "tung/reads.rds")
```

```{r}
options(stringsAsFactors = FALSE)
umi <- readRDS("tung/umi.rds")
umi.qc <- umi[rowData(umi)$use, colData(umi)$use]
endog_genes <- !rowData(umi.qc)$is_feature_control
plotPCA(
    umi[endog_genes, ],
    exprs_values = "counts",
    colour_by = "batch",
    size_by = "total_features",
    shape_by = "individual"
)

plotPCA(
    umi[endog_genes, ],
    exprs_values = "logcounts_raw",
    colour_by = "batch",
    size_by = "total_features",
    shape_by = "individual"
)
```

```{r}
plotPCA(
    umi.qc[endog_genes, ],
    exprs_values = "logcounts_raw",
    colour_by = "batch",
    size_by = "total_features",
    shape_by = "individual"
)
```

